src/cargo/core/compiler/build_config.rs

```diff
@@ -97,10 +100,23 @@ impl BuildConfig {
             build_plan: false,
             rustc_wrapper: None,
             rustfix_diagnostic_server: RefCell::new(None),
+            cache_messages: config.cli_unstable().cache_messages,
         })
     }
 
-    pub fn json_messages(&self) -> bool {
+    /// Whether or not Cargo should cache compiler messages on disk.
+    pub fn cache_messages(&self) -> bool {
+        self.cache_messages
+            && match self.message_format {
+                MessageFormat::Human | MessageFormat::Json => true,
+                // short is currently not supported
+                MessageFormat::Short => false,
```

Comment:
The goal is to eventually have `cache_messages` simply be true, right? If that's the case could combining `-Z cache-messages` and `--message-format short` be an error for now like with pipelined compilation? That'll hopefully provide us a solid reminder to fix it up before stabilizing.